### ISSUE TITLE
Path API

### ISSUE THREAD

**Issue by Utumno** [wrye-bash/wrye-bash#368]
Our Path API is showing its age:

- Instances are not that lightweight after all - it's ~~probably~~ behind the heavy use of memory by the BAIN tab
- case insensitivity blocks linux support ( #243 )
- the GPath cache/factory is (along with deprint, see #352) the main coupling of newest (library) code with bolt. We would like to see this eliminated - as we might as well at some point export the games.py or bsa_files.py modules as python libraries.
- reinventing the wheel - [pathlib](https://docs.python.org/3/library/pathlib.html) is now part of the std - [source](https://github.com/python/cpython/blob/3.6/Lib/pathlib.py). See also https://www.python.org/dev/peps/pep-0519/ for API changes to accept paths.

So:

- [ ] rework for performance (memory and speed)
- [ ] revisit cXXX properties - replace them in the related code with LowerDict APIs - cleaner, less error prone, and we erode Path so we can...
- [ ]  ...Investigate how easily we could switch to [pathlib2 2.2.1](https://pypi.python.org/pypi/pathlib2/) - _after_ we finally move to python 3. The **hard part** will be the comparisons between strings and Paths we perform implicitly (explicit is better than implicit...)

----

Remove read only

https://stackoverflow.com/a/43914402/281545

**Comment by Utumno** [comment]
Backburner for now

**Comment by Utumno** [comment]
@lojack5 this is the issue for Path refactoring. A couple notes not mentioned in the OP:

- [ ] revisit settings - we pickle Paths and that's not a good idea, we should pickle only std classes - portable, lighter etc. Especially the Installers.dat (if many loose files are present like in TESIV) takes _ages_ to load
- [ ] clearly differentiate between windows and abstract paths - in path lib those are called ["Pure paths".](https://docs.python.org/3/library/pathlib.html#pure-paths) Sometimes - but rarely - the "Pure windows path" is needed (like when writing the local save path to the game ini IIUC). We should make sure that we work with abstract paths internally and save abstract paths in settings/pickles - this way we can run bash with windows settings files on linux. We should then only instantiate Path when we need filesystem operations
- [ ] re filesystem operations - we maybe do not want to wrap those after all - rather directly using os.path (that accdepts path objects). Explicit is better than implicit, plus most of the wrapping Path does should be solved in other ways in py3 (for instance Path.open encoding is not needed any more in Py3)

**Comment by lojack5** [comment]
Yep I'm fairly familiar with `pathlib`s stuff, and its weird edge cases (like subclassing....wtf).  At least for now my plans are:

- [ ] Rename `bolt.Path` methods that have direct replacements in pathlib with those, performance be damned.
- [ ] Any non-pathlib functionality in ours that aren't easily replaced, make a (small) module that takes `PathLike` objects, rather than attaching those functionalities directly onto `bolt.Path`.  For example, `bolt.Path.walk` can be replaced by `os.walk` in Python 3 (as you've mentioned), but for an intermediate before the big changeover, will need to provide a `bolt.walk` (or at least `something.walk`, can converge on naming later), that acts like Python 3's `os.walk`, ie: on Python <3.6 call `str(obj)`, and on 3.6+ call `os.fspath(obj)` on the argument before processing.

End goal is basically, the `bolt.Path` API is a subset of `pathlib.Path`, so it can be seamlessly swapped out when the time comes.  Any additional utility functions will work with objects implementing the 
 `PathLike` interface, so it won't matter if they're the old class, the std class, or just plain strings.

Once those goals are complete, then, and only then we can start looking at performance issues (which there will be a lot, especially in BAIN).  Most of those will probably be solved with caching.  Keep in mind we're giving up memory performance in exchange for time performance on this one (ie: BAIN refresh time vs BAIN memory usage).  Really BAIN is place where `bolt.Path` performance is the bottleneck.  Bashed Patch building doesn't use enough instances for it to matter as much as BAIN.

**Comment by Utumno** [comment]
Well turns out the pathlib.Path and Path APIs differ a lot:

```
dir(Path)
['__add__', '__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '_cext', '_cs', '_ext', '_getmtime', '_onerror', '_s', '_sbody', '_setmtime', '_shead', '_sroot', '_stail', 'atime', 'backup', 'baseTempDir', 'body', 'cext', 'clearRO', 'copyTo', 'crc', 'cs', 'csbody', 'ctime', 'drive', 'editable', 'exists', 'ext', 'getNorm', 'getcwd', 'has_invalid_chars', 'head', 'headTail', 'invalid_chars_re', 'isabs', 'isdir', 'isfile', 'join', 'list', 'makedirs', 'moveTo', 'mtime', 'open', 'psize', 'relpath', 'remove', 'removedirs', 'rmtree', 'root', 's', 'sbody', 'setcwd', 'shead', 'size_mtime', 'size_mtime_ctime', 'sroot', 'stail', 'start', 'stat', 'strippedVersion', 'sys_fs_enc', 'tail', 'temp', 'tempDir', 'tempMoveTo', 'unicodeSafe', 'untemp', 'version', 'walk']
(dp := set(dir(Path))) - (dplb := set(dir(_Path)))
{'_s', 'invalid_chars_re', 'getcwd', 'isdir', 'body', 'tempMoveTo', 'getNorm', '_ext', 'shead', 'psize', 'copyTo', 'headTail', 'setcwd', 'head', 'csbody', 'tempDir', '__add__', 'atime', 'cext', 'remove', '_stail', 'cs', '_sroot', 'ext', 'tail', 'join', '__len__', 'walk', '_cs', '__setstate__', 'stail', 'makedirs', '_cext', 'crc', 's', 'rmtree', 'clearRO', '_onerror', 'temp', 'backup', 'untemp', '_getmtime', '_sbody', 'baseTempDir', 'editable', 'sroot', 'list', '__getstate__', 'size_mtime_ctime', '_shead', 'size_mtime', 'start', 'strippedVersion', 'sys_fs_enc', 'unicodeSafe', 'sbody', 'isabs', '_setmtime', 'has_invalid_chars', 'mtime', 'ctime', 'relpath', 'moveTo', 'removedirs', 'isfile', 'version'}
dplb - dp
{'as_uri', 'read_bytes', '_format_parsed_parts', 'is_absolute', 'mkdir', 'relative_to', 'owner', 'is_reserved', 'is_relative_to', 'parents', 'joinpath', '_root', '__enter__', 'is_symlink', 'read_text', '_cached_cparts', 'lstat', '_pparts', 'suffixes', 'rmdir', 'home', 'write_bytes', 'is_fifo', 'chmod', 'cwd', 'stem', '_make_child', '_init', 'absolute', 'is_mount', 'is_dir', '_accessor', 'expanduser', '__fspath__', 'is_file', '_raw_open', 'anchor', 'link_to', '_drv', '_cparts', 'parts', 'symlink_to', 'with_stem', 'write_text', 'samefile', '__exit__', '_hash', '_opener', 'iterdir', '__bytes__', 'readlink', 'group', '_from_parsed_parts', '_from_parts', '__rtruediv__', 'suffix', 'match', 'with_suffix', '__truediv__', 'name', 'resolve', 'lchmod', 'is_socket', 'is_block_device', 'as_posix', 'rglob', 'with_name', 'is_char_device', '_str', 'glob', '_parts', 'unlink', 'rename', 'replace', 'touch', '__class_getitem__', '_make_child_relpath', 'parent', '_parse_args'}

```

**Comment by Infernio** [comment]
For future reference, cause I just ran into this with my own project, this is how you subclass a pathlib path to extend it:
```python
from pathlib import Path

class MyPath(type(Path())): # ugh
    # extensions go here
```

The Path APIs all seem to work correctly like this, giving instances of your subclass back:
```
>>> from paths import MyPath
>>> list(MyPath('test').iterdir())
[MyPath('test/test1.py'), MyPath('test/test2.py'), MyPath('test/test3.py'), MyPath('test/__init__.py')]
```

**Comment by Utumno** [comment]
Wowz - yes was getting there eventually but let's merge the rest first - with FName the landscape will clear up a lot
